Add a fast-path to `Debug` ASCII `&str`

Swatinem opened the pull request:
Instead of going through the `EscapeDebug` machinery, we can just skip over ASCII chars that don’t need any escaping.

---

This is an alternative / a companion to https://github.com/rust-lang/rust/pull/121138.

The other PR is adding the fast path deep within `EscapeDebug`, whereas this skips as early as possible.

Swatinem committed:
Add a fast-path to `Debug` ASCII `&str`

Instead of going through the `EscapeDebug` machinery, we can just skip over ASCII chars that don’t need any escaping.

compiler-errors commented on library/core/src/fmt/mod.rs:
```diff
@@ -2340,6 +2340,11 @@ impl Debug for str {
         let mut from = 0;
         for (i, c) in self.char_indices() {
+            // a fast path for ASCII chars that do not need escapes:
+            if matches!(c, ' '..='~') && !matches!(c, '\\' | '\"') {
```
Do we want to use https://doc.rust-lang.org/std/primitive.char.html#method.is_ascii_graphic or some combination of other methods rather than hard-coding this?

Swatinem replied:
`is_ascii_graphic` does not *include* space, and I don’t think `\` or `"` are part of any of the other methods?

the8472 replied:
There's `escape_ascii` but it escapes more than those two.

the8472 replied:
The loop doesn't look like it would vectorize or even unroll well. For a string of chars that don't need escaping it might be worth processing the utf8 bytes directly [like EscapeAscii](https://github.com/rust-lang/rust/blob/7d3702e472b99be0f5de6608dd87af1df8f99428/library/core/src/slice/ascii.rs#L268-L280)

Swatinem commented:
While we are still bikeshedding the implementation details, I would appreciate a benchmark run. Not sure if the compiler performance testsuite will show any change, in a micro-benchmark, I was able to get a 10x improvement for a pure-ASCII string, and even a tiny improvement in the "pure"-Unicode case (because even a "pure" unicode sentence still contains ASCII spaces)

the8472 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 04602a24a38033c6ef20fe0ecb53316b808542ce with merge d3c44b181fb6776a28574e754c3db10da6172bd9...
<!-- homu: {"type":"TryBuildStarted","head_sha":"04602a24a38033c6ef20fe0ecb53316b808542ce","merge_sha":"d3c44b181fb6776a28574e754c3db10da6172bd9"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7949586729/job/21701917449)
Build commit: d3c44b181fb6776a28574e754c3db10da6172bd9 (`d3c44b181fb6776a28574e754c3db10da6172bd9`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7949586729/job/21701917449"},"merge_sha":"d3c44b181fb6776a28574e754c3db10da6172bd9"} -->

rust-timer commented:
Queued d3c44b181fb6776a28574e754c3db10da6172bd9 with parent 6f726205a1b7992537ddec96c83f2b054b03e04f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6f726205a1b7992537ddec96c83f2b054b03e04f&end=d3c44b181fb6776a28574e754c3db10da6172bd9).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (d3c44b181fb6776a28574e754c3db10da6172bd9): [comparison URL](https://perf.rust-lang.org/compare.html?start=6f726205a1b7992537ddec96c83f2b054b03e04f&end=d3c44b181fb6776a28574e754c3db10da6172bd9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6f726205a1b7992537ddec96c83f2b054b03e04f&end=d3c44b181fb6776a28574e754c3db10da6172bd9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6f726205a1b7992537ddec96c83f2b054b03e04f&end=d3c44b181fb6776a28574e754c3db10da6172bd9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.4% | [1.4%, 1.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6f726205a1b7992537ddec96c83f2b054b03e04f&end=d3c44b181fb6776a28574e754c3db10da6172bd9&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.8% | [1.8%, 1.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6f726205a1b7992537ddec96c83f2b054b03e04f&end=d3c44b181fb6776a28574e754c3db10da6172bd9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 640.535s -> 641.862s (0.21%)
**Artifact size**: 308.83 MiB -> 308.80 MiB (-0.01%)

Swatinem commented:
Not too surprisingly, the compile time benchmarks did not budge, or look bogus.

Though if you select "runtime", there is a very significant improvement to `fmt-debug-derive`, which sounds like exactly the benchmark touching this code.

https://perf.rust-lang.org/compare.html?start=6f726205a1b7992537ddec96c83f2b054b03e04f&end=d3c44b181fb6776a28574e754c3db10da6172bd9&stat=instructions%3Au&tab=runtime&nonRelevant=true&showRawData=true

---

So the question remains, what can I do to improve the confidence in this change? I have looked through other similar existing methods, but none of which do exactly what I expect.

Another option would be to also perf-test https://github.com/rust-lang/rust/pull/121138, as in my local benchmarking I found that the slowness came from binary searching the `Grapheme_Extend` table. Maybe the compiler is smart enough to generate equally fast code with an arguably simpler change?

cuviper commented:
Let's see what this now does *on top* of #121138...

@bors try @rust-timer queue

And for good measure, let's mention that #122013 is also making changes around this.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 04602a24a38033c6ef20fe0ecb53316b808542ce with merge d253c6933da1078e2c622d865089469a722c376b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"04602a24a38033c6ef20fe0ecb53316b808542ce","merge_sha":"d253c6933da1078e2c622d865089469a722c376b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8177970740/job/22364061484)
Build commit: d253c6933da1078e2c622d865089469a722c376b (`d253c6933da1078e2c622d865089469a722c376b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8177970740/job/22364061484"},"merge_sha":"d253c6933da1078e2c622d865089469a722c376b"} -->

rust-timer commented:
Queued d253c6933da1078e2c622d865089469a722c376b with parent bfe762e0ed2e95041cc12c02c5565c4368f2cc9f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&end=d253c6933da1078e2c622d865089469a722c376b).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (d253c6933da1078e2c622d865089469a722c376b): [comparison URL](https://perf.rust-lang.org/compare.html?start=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&end=d253c6933da1078e2c622d865089469a722c376b&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&end=d253c6933da1078e2c622d865089469a722c376b&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&end=d253c6933da1078e2c622d865089469a722c376b&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.5%  | [1.4%, 7.7%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-2.6%, -2.6%] | 1     |
| All ❌✅ (primary)                 | 4.5%  | [1.4%, 7.7%]   | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&end=d253c6933da1078e2c622d865089469a722c376b&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=bfe762e0ed2e95041cc12c02c5565c4368f2cc9f&end=d253c6933da1078e2c622d865089469a722c376b&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 644.65s -> 646.087s (0.22%)
**Artifact size**: 175.06 MiB -> 175.06 MiB (0.00%)

cuviper commented:
AIUI you're still investigating...
@rustbot author

Swatinem commented:
> 2 weeks ago

Yes, I would still love to get to this, just got side tracked a bit with other stuff in the meantime 😅

cuviper commented:
No problem, just managing my review queue. :)

Swatinem commented:
I was looking at this again today, and created a *tiny* micro-benchmark (might be worth adding these or similar ones to some compiler internal benchmarks which I can look into as well):

```rust
#[divan::bench]
fn dbg_ascii() {
    assert_eq!(
        format!("{:?}", "just some ascii string without escapes"),
        r#""just some ascii string without escapes""#
    );
}

#[divan::bench]
fn dbg_escapes() {
    assert_eq!(
        format!("{:?}", "some\t\"ascii'\nwith\r\\escapes"),
        r#""some\t\"ascii'\nwith\r\\escapes""#
    );
}

#[divan::bench]
fn dbg_unicode() {
    //
    assert_eq!(format!("{:?}", "Я ❤ Юля"), r#""Я ❤ Юля""#);
}
```

And that is yielding these results:

```
# 1.77.2 (no optimization)

divan           fastest       │ slowest       │ median        │ mean    
├─ dbg_ascii    924.7 ns      │ 6.156 µs      │ 956 ns        │ 962.8 ns
├─ dbg_escapes  824.7 ns      │ 7.256 µs      │ 862.2 ns      │ 867.8 ns
╰─ dbg_unicode  524.7 ns      │ 2.299 µs      │ 546.6 ns      │ 550.8 ns

# nightly (with #121138)

divan           fastest       │ slowest       │ median        │ mean    
├─ dbg_ascii    512.2 ns      │ 5.321 µs      │ 531 ns        │ 536.9 ns
├─ dbg_escapes  587.2 ns      │ 4.062 µs      │ 615.3 ns      │ 624.3 ns
╰─ dbg_unicode  499.7 ns      │ 82.89 µs      │ 599.7 ns      │ 638.9 ns

# This PR

divan           fastest       │ slowest       │ median        │ mean    
├─ dbg_ascii    446.6 ns      │ 8.34 µs       │ 468.5 ns      │ 478.5 ns
├─ dbg_escapes  743.5 ns      │ 9.193 µs      │ 768.5 ns      │ 777.7 ns
╰─ dbg_unicode  943.5 ns      │ 4.468 µs      │ 968.5 ns      │ 977.5 ns

# Using `EscapeAscii`-like skipping over ascii bytes

divan           fastest       │ slowest       │ median        │ mean    
├─ dbg_ascii    302.8 ns      │ 8.046 µs      │ 318.5 ns      │ 323.6 ns
├─ dbg_escapes  587.2 ns      │ 16.34 µs      │ 612.2 ns      │ 638.2 ns
╰─ dbg_unicode  949.7 ns      │ 5.987 µs      │ 968.5 ns      │ 979 ns  
```

Its interesting to see that this PR as it exists only improves the happy case with pure ascii.

Skipping over ascii bytes and using some `unsafe` slicing similar to the code pointed at by https://github.com/rust-lang/rust/pull/121150#discussion_r1515326716 would improve the ideal case further, as well as get the mixed case back to current levels.
However its pretty bad to see that both are still regressing the (almost) pure unicode case by quite a bit.

This does make sense, as it is now essentially running both the fast-path checks, as well as the normal code for unicode chars. I wonder if that can be avoided somehow. Wasn’t there some functionality that would split a `str` into runs of ascii-only and unicode-only chunks really quickly?

Swatinem committed:
Write `char::DebugEscape` sequences using `write_str`

Instead of writing each `char` of an escape sequence one by one,
this delegates to `Display`, which uses `write_str` internally
in order to write the whole escape sequence at once.

Swatinem commented:
I figured out that my benchmark results quoted above (vs nightly) were bogus as I compiled things with the default `library` profile, and adding the following to my `config.toml` fixed the bogus regressions:

```toml
incremental = false
lto = "thin-local"
codegen-units-std = 1
```

Now this PR is rebased on top of https://github.com/rust-lang/rust/pull/124575, and I get the following *additional* improvements:

```
 name                        write-str ns/iter  skip-ascii ns/iter  diff ns/iter   diff %  speedup 
 str::debug::ascii_escapes   1,003              735                         -268  -26.72%   x 1.36        
 str::debug::ascii_only      672                358                         -314  -46.73%   x 1.88        
 str::debug::mixed           1,549              1,491                        -58   -3.74%   x 1.04        
 str::debug::mostly_unicode  1,176              1,165                        -11   -0.94%   x 1.01        
 str::debug::some_unicode    717                439                         -278  -38.77%   x 1.63        
```

So this does already look very good. I can further micro-optimize the heck out of this, though not sure it might be worth the complexity hit.

General question for reviewers (@the8472 , feel free to steal the review if you want to): is it preferable to keep these as separate PRs and rebase them on top of each other, or rather just merge things into one PR (this one?), as I will keep things as separate commits anyway.

the8472 commented:
Imo they're small enough and closely related that a single PR should be fine.

Regarding benchmarking, see https://std-dev-guide.rust-lang.org/development/perf-benchmarking.html

Swatinem committed:
Introduce printable-ASCII fast-path for `impl Debug for str`

Instead of having a single loop that works on utf-8 `char`s,
this splits the implementation into a loop that quickly skips over
printable ASCII, falling back to per-char iteration for other chunks.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8938298562/job/24552167089) [(plain)](https://github.com/rust-lang/rust/commit/35aa3010765af3da091b33094100d7bd930d2e2f/checks/24552167089/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:0ad4b8fadaa221de15dcec353f45205ec38ea70b)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.423   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.439 Collecting boolean-py==4.0
#13 0.445   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.462 Collecting chardet==5.1.0
---
#13 3.512 Building wheels for collected packages: reuse
#13 3.513   Building wheel for reuse (pyproject.toml): started
#13 3.835   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.836   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.836   Stored in directory: /tmp/pip-ephem-wheel-cache-hxajv3mj/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.839 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.861   Attempting uninstall: setuptools
#13 3.861     Found existing installation: setuptools 59.6.0
#13 3.863     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#13 5.077   Downloading virtualenv-20.26.1-py3-none-any.whl (3.9 MB)
#13 5.211      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 29.9 MB/s eta 0:00:00
#13 5.265 Collecting filelock<4,>=3.12.2
#13 5.271   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.290 Collecting distlib<1,>=0.3.7
#13 5.296   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.307      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 52.1 MB/s eta 0:00:00
#13 5.337 Collecting platformdirs<5,>=3.9.1
#13 5.344   Downloading platformdirs-4.2.1-py3-none-any.whl (17 kB)
#13 5.428 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.588 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.1 virtualenv-20.26.1
#13 DONE 5.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      204736 kB
DirectMap2M:     8183808 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/79734f1db8dbe322192dea32c0f6b80ab14c4c1d/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-79734f1db8dbe322192dea32c0f6b80ab14c4c1d-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2396: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2397: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2404: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2406: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2411: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2418: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2420: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2424: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2431: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/fmt/mod.rs:2439: undocumented unsafe
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 46.3 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 36.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 36.0 MB/s eta 0:00:00
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 132.3 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:00:59
  local time: Fri May  3 11:36:04 UTC 2024
  network time: Fri, 03 May 2024 11:36:04 GMT
```

</details>

Swatinem commented:
I’m now on a Linux machine where I can follow the benchmarking best practices reasonably well.

I also further went into a rabbit hole, and added two more commits here:

- Splitting this up into two nested loops, handling ASCII-only and Unicode chunks separately, yielding the following perf improvements on top of the existing PR:

```
name                        skip-ascii ns/iter  separate-loops ns/iter  diff ns/iter   diff %  speedup 
 str::debug::ascii_escapes   577                 360                             -217  -37.61%   x 1.60 
 str::debug::ascii_only      383                 283                             -100  -26.11%   x 1.35 
 str::debug::mixed           1,656               1,327                           -329  -19.87%   x 1.25 
 str::debug::mostly_unicode  1,489               1,280                           -209  -14.04%   x 1.16 
 str::debug::some_unicode    444                 334                             -110  -24.77%   x 1.33 
```

- Another PR sprinkling way too much `unsafe` slicing than I would be comfortable with all over the place, for some more perf improvements:

```
name                        separate-loops ns/iter  unsafe ns/iter  diff ns/iter   diff %  speedup 
 str::debug::ascii_escapes   360                     356                       -4   -1.11%   x 1.01 
 str::debug::ascii_only      283                     232                      -51  -18.02%   x 1.22 
 str::debug::mixed           1,327                   1,254                    -73   -5.50%   x 1.06 
 str::debug::mostly_unicode  1,280                   1,193                    -87   -6.80%   x 1.07 
 str::debug::some_unicode    334                     313                      -21   -6.29%   x 1.07 
```

Even without `unsafe`, the wins from splitting this into nested loops seem to be worth the added complexity.
So I might clean that up and add a couple more comments if that approach is indeed worth following.

I’m also eager to learn if there might be some tricks to avoid `unsafe`, but achieve the same benefits, or maybe limits its use to only the places it really makes sense.

bors commented:
:umbrella: The latest upstream changes (presumably #124961) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Swatinem commented:
I rebased, reduced the number of `unsafe`, and documented the remaining uses.

@rustbot ready

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9050839666/job/24866669323) [(plain)](https://github.com/rust-lang/rust/commit/271eae93ef7bd55e2c85af82e2ee5f43aa040430/checks/24866669323/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
..................................................

failures:

---- fmt::test_format_macro_interface stdout ----
thread 'fmt::test_format_macro_interface' panicked at library/alloc/tests/fmt.rs:73:5:
assertion `left == right` failed
  left: "\"foo\\x00bar\\x01baz\\x7fqux\""
 right: "\"foo\\0bar\\u{1}baz\\u{7f}qux\""

failures:
    fmt::test_format_macro_interface

```

</details>

joboet commented on library/core/src/fmt/mod.rs:
```diff
@@ -2399,23 +2399,68 @@ impl Display for bool {
+        // which are then processed separately, to enable a fast path for the ASCII-only chunk.
+        let mut rest = self.as_bytes();
+        while rest.len() > 0 {
+            let mut ascii_bytes: &[u8];
+            let unicode_bytes: &[u8];
```
Ahhh, C-isms! 🙅 😄

joboet commented on library/core/src/fmt/mod.rs:
```diff
@@ -2399,23 +2399,68 @@ impl Display for bool {
+            // first, handle an ascii-only prefix
+            let non_ascii_position = rest.iter().position(|&b| b >= 0x80).unwrap_or(rest.len());
+            // SAFETY: the position was derived from an iterator, so is known to be within bounds, and at a char boundary
+            (ascii_bytes, rest) = unsafe { rest.split_at_unchecked(non_ascii_position) };
```
```suggestion
            let (mut ascii_bytes, rest) = unsafe { rest.split_at_unchecked(non_ascii_position) };
```

Swatinem replied:
This would shadow `rest` though :-(

joboet replied:
Ah, right!

joboet commented on library/core/src/fmt/mod.rs:
```diff
@@ -2399,23 +2399,68 @@ impl Display for bool {
+            fn needs_escape(b: u8) -> bool {
+                b > 0x7E || b < 0x20 || b == b'\\' || b == b'"'
+            }
+            while let Some(escape_position) = ascii_bytes.iter().position(|&b| needs_escape(b)) {
```
I wonder if it would be faster to combine this check with the ASCII check, and combine this loop with the Unicode loop below. I.e. find the next character that is not an ASCII printable, then escape it, check if it is easily printable, and then continue the loop.

Swatinem replied:
This makes a ton of sense, now I just skip ascii-printable prefixes, and then process a chunk of non-ascii-printable chars one by one.
I might do another iteration on the code that skips the two nested loops and rather runs the outer loop per non-ascii-printable char. Lets see how that turns out perf-wise.

Thus far I am quite surprised by whatever the optimizer is doing, as you can see in the most recent commit.

Swatinem committed:
Switch to primarily using `&str`

Surprisingly, benchmarks have shown that using `&str`
instead of `&[u8]` with some `unsafe` code is actually faster.

Swatinem committed:
Process a single not-ASCII-printable `char` per iteration

This avoids having to collect a non-ASCII-printable run before processing it.

joboet commented on library/core/src/fmt/mod.rs:
```diff
@@ -2399,23 +2399,47 @@ impl Display for bool {
+            b > 0x7E || b < 0x20 || b == b'\\' || b == b'"'
+        }
+
+        // the outer loop here splits the string into chunks of printable ASCII, which is just skipped over,
+        // and chunks of other chars (unicode, or ASCII that needs escaping), which is handler per-`char`.
```
There is no inner loop anymore, so I think this comment could use a rewrite.

joboet reviewed (approved):
This looks great and is ready to merge (modulo the comment nit), thank you!

I have some ideas on how to optimize this further (I don't think the `len_utf8` gets optimized away, even though it is redundant), but I think we can leave that for a follow-up PR...

joboet commented:
Thank you!
@bors r+

bors commented:
:pushpin: Commit 004100c222638c980b6509aba0ed4990181fa5dc has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 004100c222638c980b6509aba0ed4990181fa5dc -->
<!-- homu: {"type":"Approved","sha":"004100c222638c980b6509aba0ed4990181fa5dc","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 004100c222638c980b6509aba0ed4990181fa5dc with merge 213ad10c8f0fc275648552366275dc4e07f97462...
<!-- homu: {"type":"BuildStarted","head_sha":"004100c222638c980b6509aba0ed4990181fa5dc","merge_sha":"213ad10c8f0fc275648552366275dc4e07f97462"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9223854881/job/25383799096)
Approved by: joboet
Pushing 213ad10c8f0fc275648552366275dc4e07f97462 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"joboet","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9223854881/job/25383799096"},"merge_sha":"213ad10c8f0fc275648552366275dc4e07f97462"} -->

rust-timer commented:
Finished benchmarking commit (213ad10c8f0fc275648552366275dc4e07f97462): [comparison URL](https://perf.rust-lang.org/compare.html?start=464987730ab36e3456ba57c53965372989382deb&end=213ad10c8f0fc275648552366275dc4e07f97462&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=464987730ab36e3456ba57c53965372989382deb&end=213ad10c8f0fc275648552366275dc4e07f97462&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.5% | [-0.9%, -0.4%] | 7     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=464987730ab36e3456ba57c53965372989382deb&end=213ad10c8f0fc275648552366275dc4e07f97462&stat=max-rss)
<details>
<summary>Results (secondary 4.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 4.1% | [3.5%, 4.8%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=464987730ab36e3456ba57c53965372989382deb&end=213ad10c8f0fc275648552366275dc4e07f97462&stat=cycles:u)
<details>
<summary>Results (primary -2.4%, secondary -4.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.4% | [-3.0%, -2.0%] | 5     |
| Improvements ✅ <br /> (secondary) | -4.4% | [-8.8%, -2.4%] | 16    |
| All ❌✅ (primary)                 | -2.4% | [-3.0%, -2.0%] | 5     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=464987730ab36e3456ba57c53965372989382deb&end=213ad10c8f0fc275648552366275dc4e07f97462&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 8     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | 0.0%  | [0.0%, 0.0%]   | 1     |
</details>

**Bootstrap**: 673.01s -> 673.59s (0.09%)
**Artifact size**: 315.73 MiB -> 315.74 MiB (0.00%)
